### PR TITLE
Migrate Card, FlatCard, ImageCard, ResourceCard, and ValueCard components to typescript

### PR DESCRIPTION
**Pull request by recondesigns**
# Description of changes

*Converting components included all test and storybook files for each.*

- Converted `Card` component to Typescript, removed prop-types, and added a props type
- Converted `FlatCard` component to Typescript, removed prop-types, and added a props type
- Converted `ImageCard` component to Typescript, removed prop-types, and added a props type
- Converted `ResourceCard` component to Typescript, removed prop-types, and added a props type
    - Converted `ResourceSkeletonCard` component to Typescript, removed prop-types, and added a props type
- Converted `ValueCard` component to Typescript, removed prop-types, and added a props type

# Issue Resolved

Fixes #NA

#### All `.tsx` files for `Card`,  `FlatCard`, and `ImageCard`

Screenshot 2024-03-03 at 10 06 36 AM

#### All `.tsx` files for `ResourceCard`, `and `ValueCard`
Screenshot 2024-03-03 at 10 07 10 AM

### CODE REVIEW

File: components/Cards/ResourceCard/ResourceCard.tsx [hunk 1]
```diff
@@ -105,44 +153,58 @@ export const possibleUserVotes = {
...
-  name: string.isRequired,
-  id: number.isRequired,
-  category: string,
-  languages: oneOfType([string, array]),
-  isFree: bool,
-  handleVote: func,
-  upvotes: number,
-  userVote: oneOf(Object.values(possibleUserVotes)),
-};
+// ResourceCard.propTypes = {
```
Comment (kylemh) [inline components/Cards/ResourceCard/ResourceCard.tsx:156]: 🚧

File: components/Cards/ResourceCard/ResourceSkeletonCard.tsx [hunk 2]
```diff
@@ -1,15 +1,17 @@
...
+   */
+  numberOfSkeletons: number;
 };
 
-function ResourceSkeletonCard({ numberOfSkeletons }) {
+function ResourceSkeletonCard({ numberOfSkeletons }: ResourceSkeletonCardPropsType) {
   return (
     <div className={styles.resourcesCardWrapper}>
-      {[...Array(numberOfSkeletons)].map((skeleton, index) => (
+      {[...Array(numberOfSkeletons)].map((index) => (
```
Comment (kylemh) [inline components/Cards/ResourceCard/ResourceSkeletonCard.tsx:14]: whoops this is a mistake. we want to keep it as it was. the first argument is always undefined.

File: components/Cards/ResourceCard/__tests__/ResourceCard.test.tsx [hunk 3]
```diff
@@ -12,12 +11,13 @@ describe('ResourceCard', () => {
...
     const handleVoteMock = vi.fn();
-
     const component = render(<ResourceCard {...requiredProps} handleVote={handleVoteMock} />);
-
     const UpvoteButton = component.queryByTestId(UPVOTE_BUTTON);
+
     expect(handleVoteMock).not.toHaveBeenCalled();
-    fireEvent.click(UpvoteButton);
+
+    UpvoteButton && fireEvent.click(UpvoteButton);
```
Comment (kylemh) [inline components/Cards/ResourceCard/__tests__/ResourceCard.test.tsx:19]:
```suggestion
   fireEvent.click(UpvoteButton!);
```

we never wanna do conditional testing, but you can use the non-null assertion operator.

File: components/Cards/ResourceCard/__tests__/ResourceCard.test.tsx [hunk 4]
```diff
@@ -29,12 +29,13 @@ describe('ResourceCard', () => {
...
     const handleVoteMock = vi.fn();
-
     const component = render(<ResourceCard {...requiredProps} handleVote={handleVoteMock} />);
-
     const DownvoteButton = component.queryByTestId(DOWNVOTE_BUTTON);
+
     expect(handleVoteMock).not.toHaveBeenCalled();
-    fireEvent.click(DownvoteButton);
+
+    DownvoteButton && fireEvent.click(DownvoteButton);
```
Comment (kylemh) [inline components/Cards/ResourceCard/__tests__/ResourceCard.test.tsx:37]: same here

File: components/Cards/ResourceCard/__tests__/ResourceCard.test.tsx (hunk 3 above)
Comment (recondesigns) [reply components/Cards/ResourceCard/__tests__/ResourceCard.test.tsx:19]: Super thanks for the heads up - just got to read up on that. There is so much to types and Typescript that I am still missing out on lol.

File: components/Cards/ImageCard/ImageCard.tsx [hunk 5]
```diff
@@ -1,26 +1,35 @@
-import { bool, node, string } from 'prop-types';
+// import { bool, node, string } from 'prop-types';
```
Comment (kylemh) [inline components/Cards/ImageCard/ImageCard.tsx:1]: 🚧

File: components/Cards/ResourceCard/ResourceSkeletonCard.tsx [hunk 6]
```diff
@@ -1,14 +1,17 @@
...
+   */
+  numberOfSkeletons: number;
 };
 
-function ResourceSkeletonCard({ numberOfSkeletons }) {
+function ResourceSkeletonCard({ numberOfSkeletons }: ResourceSkeletonCardPropsType) {
   return (
     <div className={styles.resourcesCardWrapper}>
+      {/** @ts-expect-error */}
       {[...Array(numberOfSkeletons)].map((skeleton, index) => (
```
Comment (kylemh) [inline components/Cards/ResourceCard/ResourceSkeletonCard.tsx:15]:
```suggestion
      {/** @ts-expect-error */}
      {[...Array(numberOfSkeletons)].map((_skeleton, index) => (
```

does the type arrow go away if you do this?

**Review by kylemh — APPROVED** [review]

File: components/Cards/ResourceCard/ResourceSkeletonCard.tsx (hunk 6 above)
Comment (recondesigns) [reply components/Cards/ResourceCard/ResourceSkeletonCard.tsx:15]: Yea, that worked. Can you explain that, or let me know something I can search to read up on what happened there?

Screenshot 2024-03-03 at 7 50 53 PM

Comment (kylemh) [reply components/Cards/ResourceCard/ResourceSkeletonCard.tsx]: So, TypeScript complains about unused vars from a compiler standpoint beside the linter.

We have a lint rule that matches a TypeScript paradigm where prefixing unused variables with a _ means we know they're unused, but it's necessary.

The _ prefix is a thing in other languages for the same purpose too.

**Pull request merged**